lib/src/session.dart

```diff
@@ -0,0 +1,181 @@
+  void pause() => realmCore.sessionPause(this);
+
+  /// Attempts to resume the session and enable synchronization with the server.
+  /// All sessions are active by default and calling this method is only necessary
+  /// if [pause] was called previously.
+  void resume() => realmCore.sessionResume(this);
+
+  /// Waits for the [Session] to finish all pending uploads.
+  Future<void> waitForUpload() => realmCore.sessionWaitForUpload(this);
+
+  /// Waits for the [Session] to finish all pending downloads.
+  Future<void> waitForDownload() => realmCore.sessionWaitForDownload(this);
+
+  /// Gets a [Stream] of [SyncProgress] that can be used to track upload or download progress.
+  Stream<SyncProgress> getProgressStream(ProgressDirection direction, ProgressMode mode) {
+    final controller = SessionProgressNotificationsController(this, direction, mode);
+    return controller.createStream();
+  }
+}
+
+/// The current state of a [Session] object
+enum SessionState {
+  /// The session is connected to the MongoDB Realm server and is actively transferring data.
+  active,
+
+  /// The session is not currently communicating with the server.
+  inactive,
+}
+
+/// The current connection state of a [Session] object
+enum ConnectionState {
+  /// The session is disconnected from the MongoDB Realm server.
+  disconnected,
+
+  /// The session is connecting to the MongoDB Realm server.
+  connecting,
+
+  /// The session is connected to the MongoDB Realm server.
+  connected,
+}
+
+/// The transfer direction (upload or download) tracked by a given progress notification subscription.
+enum ProgressDirection {
+  /// Monitors upload progress.
+  upload,
+
+  /// Monitors download progress.
+  download
+}
+
+/// The desired behavior of a progress notification subscription.
+enum ProgressMode {
+  /// The callback will be called forever, or until it is unregistered by disposing the subscription token.
+  /// Notifications will always report the latest number of transferred bytes, and the most up-to-date number of
+  /// total transferable bytes.
+  reportIndefinitely,
+
+  /// The callback will, upon registration, store the total number of bytes to be transferred. When invoked, it will
+  /// always report the most up-to-date number of transferable bytes out of that original number of transferable bytes.
+  /// When the number of transferred bytes reaches or exceeds the number of transferable bytes, the callback will
+  /// be unregistered.
+  forCurrentlyOutstandingWork,
+}
+
+/// A type containing information about the progress state at a given instant.
+class SyncProgress {
+  /// The number of bytes that have been transferred since subscribing for progress notifications.
+  final int transferredBytes;
+
+  /// The total number of bytes that have to be transferred since subscribing for progress notifications.
+  /// The difference between that number and [transferredBytes] gives you the number of bytes not yet
```

Comment:
Because it's rarely useful to show the difference. The bytes reported are very rough estimates and can't be used for anything more meaningful than showing a progress bar to the user. Those typically look like `12.4 MB / 22.9 MB` so the difference is not useful there. Users may also want to show a percentage in which case they'd need to subtract the two values, but both are operations they can do themselves.